onnxscript/optimizer/_constant_folding.py

```diff
@@ -0,0 +1,712 @@
+    constraint in that direction.
+    """
+
+    min_version: int | None
+    max_version: int | None
+    function: PartialEvaluatorFunction
+
+    def valid_for(self, version: int) -> bool:
+        """Returns True if this evaluator is applicable for the given version."""
+        return (self.min_version is None or version >= self.min_version) and (
+            self.max_version is None or version <= self.max_version
+        )
+
+
+class PartialEvaluatorRegistry:
+    """A class that maintains a registry of evaluators for ops."""
+
+    def __init__(self):
+        self.op_evaluators: dict[tuple[str, str], list[PartialEvaluator]] = {}
+
+    def lookup_evaluators(self, domain: str, opname: str, version: int):
+        evaluator_list = self.op_evaluators.get((domain, opname), [])
+        return [
+            evaluator.function for evaluator in evaluator_list if evaluator.valid_for(version)
+        ]
+
+    def register(self, opname: str, domain: str = "", version=None):
+        if (domain, opname) not in self.op_evaluators:
+            evaluator_list = []
+            self.op_evaluators[(domain, opname)] = evaluator_list
+        else:
+            evaluator_list = self.op_evaluators[(domain, opname)]
+        if version is None:
+            min_version = None
+            max_version = None
+        elif isinstance(version, int):
+            min_version = version
+            max_version = version
+        elif isinstance(version, tuple):
+            min_version, max_version = version
+
+        def decorator(function: PartialEvaluatorFunction) -> PartialEvaluatorFunction:
+            evaluator_list.append(PartialEvaluator(min_version, max_version, function))
+            return function
+
+        return decorator
+
+
+registry: PartialEvaluatorRegistry = PartialEvaluatorRegistry()
+
+register = registry.register
+
+
+def _get_numpy_value(val: ir.Value) -> np.ndarray | None:
+    const_value = val.const_value
+    if hasattr(const_value, "numpy"):
+        return const_value.numpy()
+    return None
+
+
+def _get_bool_value(val: ir.Value | None) -> bool | None:
+    if val is None:
+        return None
+    val = _get_numpy_value(val)
+    if val is None:
+        return None
+    if isinstance(val, bool):
+        return val
+    if isinstance(val, np.bool_):
+        return bool(val)
+    if isinstance(val, np.ndarray) and val.size == 1 and val.dtype == bool:
+        return val.item(0)
+    return None
+
+
+def _get_input(node: ir.Node, index: int) -> ir.Value | None:
+    if index < len(node.inputs):
+        return node.inputs[index]
+    return None
+
+
+def _get_output(node: ir.Node, index: int) -> ir.Value | None:
+    if index < len(node.outputs):
+        return node.outputs[index]
+    return None
+
+
+def _update_type(value: ir.Value, type: ir.TypeProtocol) -> None:
+    # TODO: merge types
+    value.type = type
+
+
+def _get_input_element_type(node: ir.Node, index: int) -> int:
```

Comment:
```suggestion
def _get_input_element_type(node: ir.Node, index: int) -> ir.DataType:
```

seems appropriate based on usage

Reply:
skipping this for now. (will require dropping the `.value` in the implementation in two places as well.)